content/Hash functions in daily life I - Martin Rubi.md

```diff
@@ -0,0 +1,172 @@
+# Hash functions in daily life - part I
+
+I want to talk about passwords. Like the ones used to log into email accounts.
+
+Hence I will talk a bit about ... laundry
+
+## Doing the laundry with a hash function
+
+Many years ago I moved to a new neighborhood and I used to do the laundry in a
+store nearby.
+
+I used to give the order number to the cashier and she would look upon all the
+laundry bags for the one with the matching number.
+
+The store was quite crowded and handled a lot of orders. Picking the laundry was
+a pretty slow process.
+
+During the time it took to find the laundry bag I used to think about how the search
+could be improved.
+
+I'll propose a possible optimization in the sections that follow.
+
+## The orders grid
+
+Each laundry bag was kept in a shelf made of boxes, like the one below
+
+
+                  -------------------
+                  |  |  |  |  |  |  |
+                  -------------------
+                  |  |  |  |  |  |  |
+                  -------------------
+                  |  |  |  |  |  |  |
+                  -------------------
+
+Each box could contain 0 or more bags.
+
+The grid was filled by choosing the first empty space and storing the laundry bag
+in it.
+
+                  -------------------------
+                  |   |   |   |   |   |   |
+                  -------------------------
+                  | . | . | . |   |   |   |
+                  -------------------------
+                  | . | . | . | . | . | . |
+                  -------------------------
+
+
+The first empty space was either an empty box or a partially filled box with
+enough space for another bag.
+
+If the grid had no space left the bags were stacked aside.
+
+                  -------------------------
+                  | . | . | . | . | . | . |
+                  -------------------------
+                  | . | . | . | . | . | . |
+                  -------------------------
+                  | . | . | . | . | . | . |
+                  -------------------------   ...
+
+
+To find the bag with a matching order number the cashier performed a lookup upon
+all the bags in the grid.
+
+That's a linear lookup. It's O(N), meaning that, let N be the number of
+laundry bags stored in the grid, the lookup would take an average of N comparisons
+to find the searched one.
+
+Actually it would take an average of N/2 comparisons but the order of magnitude
+between N and N/2 is the same and it´s said to be O(N) for simplicity.
```

Comment:
```suggestion
Actually it would take an average of N/2 comparisons but the order of magnitude
between N and N/2 is the same and it's said to be O(N) for simplicity.
```